src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSource.cs

```diff
@@ -593,20 +630,43 @@ private static SuggestedActionSetPriority GetSuggestedActionSetPriority(CodeActi
                 SnapshotSpan range,
                 CancellationToken cancellationToken)
             {
-                // Explicitly hold onto below fields in locals and use these locals throughout this code path to avoid crashes
-                // if these fields happen to be cleared by Dispose() below. This is required since this code path involves
-                // code that can run asynchronously from background thread.
-                var view = _textView;
-                var buffer = _subjectBuffer;
                 var provider = _owner;
 
-                if (view == null || buffer == null || provider == null)
+                if (IsDisposed)
                 {
+                    // We've already been disposed.  No point in continuing.
                     return false;
                 }
 
-                using (var asyncToken = provider.OperationListener.BeginAsyncOperation("HasSuggestedActionsAsync"))
+                using (var asyncToken = _owner.OperationListener.BeginAsyncOperation("HasSuggestedActionsAsync"))
                 {
+                    // Acquire the user's selection up front, before we do any async work, so that we can
+                    // directly grab it when we're on the UI thread.  That way we don't have any reentrancy
+                    // blocking concerns if VS wants to block on this call (for example, if the user 
+                    // explicitly invokes the 'show smart tag' command).
+                    //
+                    // This work must happen on the UI thread as it needs to access the _textView's mutable
+                    // state.
+                    TextSpan? selection = null;
+                    if (IsForeground())
+                    {
+                        selection = TryGetCodeRefactoringSelection(range);
+                    }
+                    else
```

Comment:
The background call won't be coming from the one scenario that led to this PR, since that will be fixed to call from the UI thread. But to answer your question, I think the answer is "you don't know. And yes, it can lead to more deadlocks."

The only way to do this without deadlocking in those scenarios is to (ONE of):
1. Require the caller to be on the UI thread (and leverage the SynchronizationContext to get you back to it if you need it).
2. Use the JoinableTaskFactory to switch to the UI thread. This lets you choose below input priority as well, if you want. Although typically it's the scenario owner that sets the priority for you so you don't usually have to do anything special there.